### PR TITLE
don't update currency rates on transient errors

### PR DESCRIPTION
**Pull request by bergeron**
## Explanation

We think in https://github.com/MetaMask/metamask-mobile/issues/11035, the intermittent 0 balances were caused by transient errors in requests to crypto compare.  Instead of setting the conversion rate to null in this case, we'll omit the state update and keep the conversion rate from the previous successful fetch.

### `@metamask/assets-controllers`

- **FIXED**: In `CurrencyRateController` if unexpected errors occur during requests to crypto compare, the conversion rate in state will remain unchanged instead of being set to null.

## Checklist

- [ ] I've updated the test suite for new or updated code as appropriate
- [ ] I've updated documentation (JSDoc, Markdown, etc.) for new or updated code as appropriate
- [ ] I've highlighted breaking changes using the "BREAKING" category above as appropriate

### CODE REVIEW

File: packages/assets-controllers/src/CurrencyRateController.ts [hunk 1]
```diff
@@ -196,23 +197,27 @@ export class CurrencyRateController extends StaticIntervalPollingController<
           error.message.includes('market does not exist for this coin pair')
         )
       ) {
+        // Don't update state on transient / unexpected errors
+        shouldUpdateState = false;
         throw error;
       }
     } finally {
       try {
```
Comment (tommasini) [inline packages/assets-controllers/src/CurrencyRateController.ts:205]: Just out of curiosity, what is the reason for this try - finally block inside of the finally block?

Comment (bergeron) [reply packages/assets-controllers/src/CurrencyRateController.ts:205]: to be sure the mutex always gets released

File: packages/assets-controllers/src/CurrencyRateController.ts [hunk 2]
```diff
@@ -196,23 +197,27 @@ export class CurrencyRateController extends StaticIntervalPollingController<
...
-        this.update(() => {
-          return {
-            currencyRates: {
-              ...currencyRates,
-              [nativeCurrency]: {
-                conversionDate,
-                conversionRate,
-                usdConversionRate,
+        if (shouldUpdateState) {
```
Comment (tommasini) [inline packages/assets-controllers/src/CurrencyRateController.ts:206]: Just wondering if it wouldn't be a more common pattern, if we just want to update the state when an error is not thrown, we can update the state as the last line of the try block, and have the finally just to release the mutex

Comment (bergeron) [reply packages/assets-controllers/src/CurrencyRateController.ts:206]: we want the state updated on some kids of errors though.  If we get `market does not exist for this coin pair`, then we want to update it to null.  Trying to maintain that existing behavior.

File: packages/assets-controllers/src/CurrencyRateController.ts (hunk 1 above)
Comment (tommasini) [reply packages/assets-controllers/src/CurrencyRateController.ts:205]: Yeah, I was wondering if the [update function](https://github.com/MetaMask/core/blob/b9e983011b6bbec73d8ae0e38921f45ab19ae782/packages/base-controller/src/BaseControllerV2.ts#L188), could indeed throw an error

**Review by gambinish — APPROVED** [review]

File: packages/assets-controllers/src/CurrencyRateController.test.ts [hunk 3]
```diff
@@ -478,4 +478,35 @@ describe('CurrencyRateController', () => {
 
     controller.destroy();
   });
+
+  it('should not update state on unexpected / transient errors', async () => {
+    const cryptoCompareHost = 'https://min-api.cryptocompare.com';
+    nock(cryptoCompareHost)
```
Comment (gambinish) [inline packages/assets-controllers/src/CurrencyRateController.test.ts:484]: nock?

File: packages/assets-controllers/src/CurrencyRateController.ts (hunk 2 above)
Comment (tommasini) [reply packages/assets-controllers/src/CurrencyRateController.ts:206]: Ah I see! Thanks for the explanation!

**Comment by tommasini** [comment]
Wondering your thoughts here: https://github.com/MetaMask/metamask-mobile/pull/11087#issuecomment-2334772379

File: packages/assets-controllers/src/CurrencyRateController.test.ts (hunk 3 above)
Comment (bergeron) [reply packages/assets-controllers/src/CurrencyRateController.test.ts:484]: https://github.com/nock/nock

**Review by tommasini — APPROVED** [review]

**Pull request merged**